File: backend/src/main/java/com/altstay/api/property/Property.java
package com.altstay.api.property;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalTime;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "property")
@Getter
@Setter
@NoArgsConstructor
public class Property {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @Column(name = "tenant_id", nullable = false)
    private UUID tenantId;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "slug", nullable = false)
    private String slug;

    @Column(name = "legal_name")
    private String legalName;

    @Column(name = "description")
    private String description;

    @Column(name = "status", nullable = false)
    private String status = "ACTIVE";

    @Column(name = "timezone", nullable = false)
    private String timezone;

    @org.hibernate.annotations.JdbcTypeCode(java.sql.Types.CHAR)
    @Column(name = "currency_code", nullable = false, length = 3)
    private String currencyCode;

    @org.hibernate.annotations.JdbcTypeCode(java.sql.Types.CHAR)
    @Column(name = "country_code", length = 2)
    private String countryCode;

    @Column(name = "address_line1")
    private String addressLine1;

    @Column(name = "address_line2")
    private String addressLine2;

    @Column(name = "city")
    private String city;

    @Column(name = "state_region")
    private String stateRegion;

    @Column(name = "postal_code")
    private String postalCode;

    @Column(name = "contact_email")
    private String contactEmail;

    @Column(name = "contact_phone")
    private String contactPhone;

    @Column(name = "check_in_time", nullable = false)
    private LocalTime checkInTime = LocalTime.of(14, 0);

    @Column(name = "check_out_time", nullable = false)
    private LocalTime checkOutTime = LocalTime.of(11, 0);

    @Column(name = "tax_rate_bps", nullable = false)
    private Integer taxRateBps = 0;

    @Column(name = "created_at", nullable = false, insertable = false, updatable = false)
    private OffsetDateTime createdAt;

    /**
     * There is deliberately no constructor without a timezone and currency. Both are {@code not
     * null} in the schema and neither has a default anywhere in this system (§2) — a convenience
     * overload that filled them in was how a property could be created with an invented timezone.
     */
    public Property(UUID tenantId, String name, String slug, String timezone, String currencyCode) {
        this.tenantId = tenantId;
        this.name = name;
        this.slug = slug;
        this.timezone = timezone;
        this.currencyCode = currencyCode;
        this.status = "ACTIVE";
        this.checkInTime = LocalTime.of(14, 0);
        this.checkOutTime = LocalTime.of(11, 0);
        this.taxRateBps = 0;
    }
}

File: frontend/src/lib/contracts/property.ts
import { z } from 'zod';

// Mirrors com.altstay.api.property.PropertyController.PropertyResponse. Twenty-two fields;
// `id` and `tenantId` are UUIDs, `checkInTime`/`checkOutTime` are LocalTime, `createdAt` is
// OffsetDateTime, `amenities` is a flat list of amenity codes.
export const PropertyResponseSchema = z.object({
  id: z.string().uuid(),
  tenantId: z.string().uuid(),
  name: z.string(),
  slug: z.string(),
  legalName: z.string().nullable(),
  description: z.string().nullable(),
  status: z.string(),
  timezone: z.string(),
  currencyCode: z.string().length(3),
  countryCode: z.string().length(2).nullable(),
  addressLine1: z.string().nullable(),
  addressLine2: z.string().nullable(),
  city: z.string().nullable(),
  stateRegion: z.string().nullable(),
  postalCode: z.string().nullable(),
  contactEmail: z.string().nullable(),
  contactPhone: z.string().nullable(),
  checkInTime: z.string().regex(/^\d{2}:\d{2}(:\d{2})?$/),
  checkOutTime: z.string().regex(/^\d{2}:\d{2}(:\d{2})?$/),
  taxRateBps: z.number().int(),
  amenities: z.array(z.string()),
  createdAt: z.string().datetime({ offset: true }),
});
export type PropertyResponse = z.infer<typeof PropertyResponseSchema>;

// Mirrors com.altstay.api.property.PropertyController.UpdatePropertyRequest. No `slug` — it's
// read-only after creation and isn't part of this request at all. `PUT` takes the full request,
// so the settings form always submits every field, including ones the user didn't touch
// (load-modify-save, not a patch — phase-6 §4.7). Only `taxRateBps` carries a real backend
// constraint (`@Min(0) @Max(10000)`); the rest is boundary validation this BFF adds on its own,
// since the Java DTO itself has no `@Valid` beyond that one field.
export const UpdatePropertyRequestSchema = z.object({
  name: z.string().min(1, 'name is required'),
  legalName: z.string().nullable(),
  description: z.string().nullable(),
  status: z.string().min(1, 'status is required'),
  timezone: z.string().min(1, 'timezone is required'),
  currencyCode: z.string().length(3),
  countryCode: z.string().length(2).nullable(),
  addressLine1: z.string().nullable(),
  addressLine2: z.string().nullable(),
  city: z.string().nullable(),
  stateRegion: z.string().nullable(),
  postalCode: z.string().nullable(),
  contactEmail: z.string().nullable(),
  contactPhone: z.string().nullable(),
  checkInTime: z.string().regex(/^\d{2}:\d{2}(:\d{2})?$/),
  checkOutTime: z.string().regex(/^\d{2}:\d{2}(:\d{2})?$/),
  taxRateBps: z.number().int().min(0).max(10000),
  amenities: z.array(z.string()),
});
export type UpdatePropertyRequest = z.infer<typeof UpdatePropertyRequestSchema>;
